autobahn/twisted/test/test_websocket_agent.py

```diff
@@ -0,0 +1,93 @@
+
+from twisted.trial import unittest
+from twisted.internet.defer import inlineCallbacks, Deferred
+from autobahn.twisted.testing import create_memory_agent
+from autobahn.twisted.websocket import WebSocketServerProtocol
+from autobahn.twisted.websocket import WebSocketClientProtocol
+
+from twisted.internet import reactor
+
+
+class SpammingWebSocketServerProtocol(WebSocketServerProtocol):
+    def onMessage(self, *args, **kw):
+        print("SERVER MESSAGE: {} {}".format(args, kw))
+
+
+class LoggingWebSocketServerProtocol(WebSocketServerProtocol):
+
+    messages = []
+    reactor = None
+
+    def onOpen(self):
+        # "break the loop" in synchonous agent behavior that means
+        # we'll "do" all the message-sending before anything gets a
+        # chance to listn ... but XXX THINK do we actually need to do
+        # this "in autobahn" or is this a produce of "all in-memory
+        # transports" which take zero time?
+        def send_messages():
+            for msg in self.messages:
+                print("sending {}".format(msg))
+                self.sendMessage(msg)
+        self.reactor.callLater(0, send_messages)
+
+    def onMessage(self, *args, **kw):
+        print("SERVER GOT MESSAGE: {} {}".format(args, kw))
+
+
+class TestAgent(unittest.TestCase):
+
+    def setUp(self):
+        pass
+
+    @inlineCallbacks
+    def test_foo(self):
+        agent = create_memory_agent(SpammingWebSocketServerProtocol, None)
+        proto = yield agent.open("ws://localhost:1234/ws", dict())
```

Comment:
I guess the path component of the URI is ignored here.  It might make sense to respect it at some point.  Requiring that the URI be completely correct for the agent to actually find a protocol is a really convenient way to have your tests prove you've gotten your URI handling correct.  Otherwise you want a separate set of tests for that.